src/security/KeyData.cc

```diff
@@ -97,33 +98,36 @@ Security::KeyData::loadX509ChainFromFile()
         return;
     }
 
-#if TLS_CHAIN_NO_SELFSIGNED // ignore self-signed certs in the chain
     if (SelfSigned(*cert)) {
-        debugs(83, DBG_PARSE_NOTE(2), "Certificate is self-signed, will not be chained: " << *cert);
-    } else
-#endif
-    {
+        debugs(83, DBG_PARSE_NOTE(2), "Signing certificate is self-signed: " << *cert);
+        // TODO: Warn if there are other (unusable) certificates present.
+    } else {
         debugs(83, DBG_PARSE_NOTE(3), "Using certificate chain in " << certFile);
         // and add to the chain any other certificate exist in the file
         CertPointer latestCert = cert;
 
+        // XXX: The first ca value is usually not a CA certificate because we
+        // loop from the very first certificate in certFilename, and that
+        // certificate is a copy of the already loaded _signing_ this->cert.
         while (const auto ca = Ssl::ReadOptionalCertificate(bio)) {
```

Comment:
@chtsanti has already written the code to address this and most other (possibly all) XXXs and TODOs added in this PR.  I did not add those improvements to this bug-fixing PR because the associated refactoring noise completely hides the bug fix. I think those refactoring changes are best reviewed and merged in a dedicated PR, separate from this "surgical" bug fix.

If you insist on this or any other PR-added TODOs and XXXs to be removed/excluded, just say so. No need to justify your change request -- to save time, I will simply remove each controversial comment (including all of them) from this PR without arguing.

Reply:
This XXX implies that we are going to begin the practice of lying to admin about what is going on. Squid is currently very clear when it loads that first certificate that it is skipped due to not being the issuer of itself - except when it is.